cargo-concordium/src/build.rs

```diff
@@ -93,15 +101,244 @@ impl SchemaBuildOptions {
+    output_wasm: Vec<u8>,
+    /// Information about the build so that it can be reproduced.
+    build_info:  utils::VersionedBuildInfo,
+    /// The sources that were built, archived as a tar file.
+    tar_archive: TarArchiveData,
+}
+
+fn build_in_container<'a>(
+    image: String,
+    package_target_dir: &Path,
+    package_root_path: &Path,
+    extra_args: impl Iterator<Item = &'a String>,
+    container_runtime: &str,
+    out_path: &Path, // canonical, fully expanded path
+    tar_path: &Path, // canonical, fully expanded path
+) -> anyhow::Result<ContainerBuildOutput> {
+    let tar_archive = create_archive(package_root_path, &[out_path, tar_path, package_target_dir])?;
+
+    let archive_hash = sha2::Sha256::digest(&tar_archive.tar_archive);
+
+    let build_command = [
+        "cargo",
+        "--locked",
+        "build",
+        "--target",
+        "wasm32-unknown-unknown",
+        "--release",
+        "--target-dir",
+        "/b/t",
+    ]
+    .into_iter()
+    .map(String::from)
+    .chain(extra_args.cloned())
+    .collect::<Vec<String>>();
+
+    // If both the potential output files exist check if there is no point
+    // rebuilding.
+    let mut output_wasm = Vec::new();
+    let mut built_alread = false;
+
+    // Check if we have to rebuild.
+    'skip: {
+        if out_path.try_exists()? && tar_path.try_exists()? {
+            let stored_tar_archive = std::fs::read(tar_path).context("Unable to read archive")?;
+            if tar_archive.tar_archive != stored_tar_archive {
+                break 'skip;
+            }
+            let stored_source =
+                WasmModule::from_file(out_path).context("Unable to read output file.")?;
+            let mut skeleton = parse_skeleton(stored_source.source.as_ref())
+                .context("Unable to parse stored output file.")?;
+            let Ok(build_info) = utils::get_build_info_from_skeleton(&skeleton) else {
+                break 'skip;
+            };
+            let utils::VersionedBuildInfo::V0(build_info) = build_info;
+            // If the sources are the same, and those sources were built with the same
+            // command in the same image then we don't need to rebuild.
+            if build_info.archive_hash.as_ref() == &archive_hash[..]
+                && build_info.build_command == build_command
+                && build_info.image == image
+            {
+                strip(&mut skeleton);
+                skeleton.output(&mut output_wasm)?;
+                built_alread = true;
+            }
+        }
+    }
+
+    if !built_alread {
+        output_wasm = build_archive(
+            &image,
+            &tar_archive.tar_archive,
+            container_runtime,
+            &build_command,
+        )
+        .context("Unable to build.")?;
+    }
+
+    let build_info = utils::VersionedBuildInfo::V0(utils::BuildInfo {
+        archive_hash: <[u8; 32]>::from(archive_hash).into(),
+        source_link: None, // TODO
```

Comment:
Todo

Reply:
I now take an optional argument to cargo build that is used here.